File: Terminal/TimeWords.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace Hash.Terminal
{
    /// <summary>
    /// The clock vocabulary for a time-of-day argument, and the conversion from a word or a clock reading to
    /// the console's own <c>hhmm</c> token.
    ///
    /// <para>A time argument is the one slot where the model has nothing to work with. Over the eleven
    /// <c>settime</c> cases in the benchmark it answers <c>8</c> for "8am" and <c>0</c> for noon, in every
    /// language and with or without the tuned adapter - it reads the hour correctly and then has no way to
    /// know that this game counts time as 800 and that noon is 1200. Neither is a fact about language, so
    /// neither is something to train: the hour-to-hhmm step is arithmetic, and noon is a value this mod
    /// knows and the model cannot.</para>
    ///
    /// <para>So the slot is declared as a word the model can pick - the same shape that already works for
    /// items and weather, where offering the whole short list of values raised recall from 32 of 53 to 40 -
    /// and every answer, word or number, is converted here. Rule-based normalisation over a recognised
    /// expression is what HeidelTime and SUTime have always done for this, and a fixed target format is
    /// exactly the case where it still beats asking a model to produce the number itself.</para>
    /// </summary>
    internal static class TimeWords
    {
        /// <summary>The words offered whatever the player wrote, so the enum is never empty.</summary>
        private static readonly string[] CoreWords =
            { "midnight", "dawn", "morning", "noon", "afternoon", "evening", "night" };

        /// <summary>How many words one slot offers. The vocabulary is fixed and short, unlike a live catalogue.</summary>
        private const int MaxWords = 12;

        /// <summary>
        /// Every word this understands, folded, with the time it means.
        ///
        /// The hours are clock conventions rather than game events: a player who asks for "evening" wants a
        /// rough time of day, and any hour inside the band answers them. Only noon, midnight and the game's
        /// own "day" have one right answer.
        /// </summary>
        private static readonly Dictionary<string, int> Table = new(StringComparer.Ordinal)
        {
            // English
            ["midnight"] = 0, ["dawn"] = 600, ["sunrise"] = 600, ["morning"] = 800, ["noon"] = 1200,
            ["midday"] = 1200, ["day"] = 1200, ["daytime"] = 1200, ["afternoon"] = 1500,
            ["evening"] = 1900, ["sunset"] = 2000, ["dusk"] = 2000, ["night"] = 2300,
            // German
            ["mitternacht"] = 0, ["morgengrauen"] = 600, ["sonnenaufgang"] = 600, ["morgens"] = 800,
            ["morgen"] = 800, ["vormittag"] = 1000, ["mittag"] = 1200, ["tag"] = 1200,
            ["nachmittag"] = 1500, ["abend"] = 1900, ["abends"] = 1900, ["sonnenuntergang"] = 2000,
            ["nacht"] = 2300, ["nachts"] = 2300,
            // Spanish
            ["medianoche"] = 0, ["amanecer"] = 600, ["manana"] = 800, ["mediodia"] = 1200, ["dia"] = 1200,
            ["tarde"] = 1500, ["atardecer"] = 2000, ["anochecer"] = 2000, ["noche"] = 2300,
            // French
            ["minuit"] = 0, ["aube"] = 600, ["matin"] = 800, ["midi"] = 1200, ["jour"] = 1200,
            ["journee"] = 1200, ["apresmidi"] = 1500, ["soir"] = 1900, ["soiree"] = 1900,
            ["coucherdusoleil"] = 2000, ["nuit"] = 2300,
        };

        /// <summary>True when this argument label names a time of day.</summary>
        internal static bool Owns(string label) =>
            !string.IsNullOrEmpty(label) && label.IndexOf("hhmm", StringComparison.OrdinalIgnoreCase) >= 0;

        /// <summary>
        /// What to offer for this request: a clock reading the player wrote, then the words their wording
        /// contains, then the core.
        ///
        /// A German player writes "mittag" and needs that word on the list, but forty words across four
        /// languages is a list nobody reads. Matching by containment is enough for a fixed vocabulary - the
        /// fuzzy ranking that a thousand-item catalogue needs would only put "tag" on an English list.
        /// </summary>
        internal static IReadOnlyList<string> Choices(string query)
        {
            var chosen = new List<string>();
            string folded = Fold(query ?? "");

            // A clock reading the player wrote goes on the list first, as the reading the console wants.
            // Without it the word list pulls the answer away from a number that was there all along:
            // offered midnight and dawn beside nothing else, "set the time to 8am" is answered midnight.
            foreach (string reading in Readings(query))
            {
                if (chosen.Count >= MaxWords) break;
                if (!chosen.Contains(reading)) chosen.Add(reading);
            }

            if (folded.Length > 0)
            {
                foreach (KeyValuePair<string, int> entry in Table)
                {
                    if (chosen.Count >= MaxWords) break;
                    if (folded.IndexOf(entry.Key, StringComparison.Ordinal) >= 0 && !chosen.Contains(entry.Key))
                        chosen.Add(entry.Key);
                }
            }

            foreach (string word in CoreWords)
            {
                if (chosen.Count >= MaxWords) break;
                if (!chosen.Contains(word)) chosen.Add(word);
            }

            return chosen;
        }

        /// <summary>
        /// Every console reading the player's own words carry - "8am", "8 uhr", "20:00".
        ///
        /// Read in pairs as well as singly, because the hour and its unit are usually two words.
        /// </summary>
        private static IEnumerable<string> Readings(string query)
        {
            string[] parts = (query ?? "").Split((char[])null, StringSplitOptions.RemoveEmptyEntries);
            for (int at = 0; at < parts.Length; at++)
            {
                if (TryToken(parts[at], out string one) && Fold(parts[at]).Length > 0
                    && !Table.ContainsKey(Fold(parts[at])))
                    yield return one;
                if (at + 1 < parts.Length && TryToken(parts[at] + " " + parts[at + 1], out string two)
                    && !Table.ContainsKey(Fold(parts[at] + parts[at + 1])))
                    yield return two;
            }
        }

        /// <summary>
        /// The console token for a word or a clock reading, or false when this is not a time at all.
        ///
        /// Accepts what the model and the player actually produce: <c>noon</c>, <c>8</c>, <c>8am</c>,
        /// <c>8h30</c>, <c>20:00</c> and the game's own <c>1530</c>. A bare hour becomes the full reading,
        /// which is the conversion the model never makes on its own.
        /// </summary>
        internal static bool TryToken(string value, out string token)
        {
            token = null;
            string text = (value ?? "").Trim();
            if (text.Length == 0) return false;

            if (TryClock(text, out int reading))
            {
                token = reading.ToString(CultureInfo.InvariantCulture);
                return true;
            }

            if (Table.TryGetValue(Fold(text), out int known))
            {
                token = known.ToString(CultureInfo.InvariantCulture);
                return true;
            }

            return false;
        }

        /// <summary>
        /// Read a clock expression: digits, then an optional separator and minutes, then an optional am/pm.
        ///
        /// Three or four digits with no separator are the console's own form and stay as they are, so
        /// <c>1530</c> is half past three and not fifteen hundred hours past midnight.
        /// </summary>
        private static bool TryClock(string text, out int reading)
        {
            reading = 0;
            int at = 0;

            int hours = 0, digits = 0;
            while (at < text.Length && char.IsDigit(text[at]) && digits < 4)
            {
                hours = (hours * 10) + (text[at] - '0');
                digits++;
                at++;
            }

            if (digits == 0) return false;

            int minutes = 0;
            bool split = false;

            if (at < text.Length && (text[at] == ':' || text[at] == '.' || text[at] == 'h' || text[at] == 'H'))
            {
                int separator = at++;
                int minuteDigits = 0;
                while (at < text.Length && char.IsDigit(text[at]) && minuteDigits < 2)
                {
                    minutes = (minutes * 10) + (text[at] - '0');
                    minuteDigits++;
                    at++;
                }

                // "8h" is eight o'clock; "8:" is a stray colon and the reading stops being one.
                if (minuteDigits == 0 && text[separator] == ':') return false;
                split = true;
            }
            else if (digits >= 3)
            {
                minutes = hours % 100;
                hours /= 100;
                split = true;
            }

            string suffix = Fold(text.Substring(at));
            if (suffix == "pm" && hours < 12) hours += 12;
            else if (suffix == "am" && hours == 12) hours = 0;
            else if (suffix.Length > 0 && suffix != "am" && suffix != "pm" && suffix != "uhr" && suffix != "h")
                return false;

            if (hours > 24 || minutes > 59) return false;
            if (hours == 24 && minutes > 0) return false;
            if (!split && digits > 2) return false;

            reading = ((hours % 24) * 100) + minutes;
            return true;
        }

        /// <summary>
        /// Letters and digits, lower case, without accents.
        ///
        /// Written out rather than taken from <c>string.Normalize</c>, which needs the globalization data
        /// MelonLoader starts the runtime without.
        /// </summary>
        private static string Fold(string value)
        {
            var sb = new StringBuilder();
            foreach (char raw in value ?? "")
            {
                char c = char.ToLowerInvariant(raw);
                switch (c)
                {
                    case 'á': case 'à': case 'â': case 'ä': case 'ã': case 'å': c = 'a'; break;
                    case 'é': case 'è': case 'ê': case 'ë': c = 'e'; break;
                    case 'í': case 'ì': case 'î': case 'ï': c = 'i'; break;
                    case 'ó': case 'ò': case 'ô': case 'ö': case 'õ': c = 'o'; break;
                    case 'ú': case 'ù': case 'û': case 'ü': c = 'u'; break;
                    case 'ñ': c = 'n'; break;
                    case 'ç': c = 'c'; break;
                    case 'ß': sb.Append("ss"); continue;
                }

                if (char.IsLetterOrDigit(c)) sb.Append(c);
            }

            return sb.ToString();
        }
    }
}

File: Tools/NeedleTraining/time_words.py
"""The clock vocabulary, mirroring Terminal/TimeWords.cs.

Two sides have to agree about a time slot or the measurement is about the mismatch rather than about the
model: the corpus and the benchmark render the schema here, and the mod resolves the answer there. Keep
the table, the folding and the clock reader identical - a word this file knows and the mod does not is a
case that scores right and fails in the game.

    from time_words import owns, choices, token, word_for, recoverable
"""

from __future__ import annotations

# The words offered whatever the player wrote, so the enum is never empty.
CORE = ["midnight", "dawn", "morning", "noon", "afternoon", "evening", "night"]

# How many words one slot offers. The vocabulary is fixed and short, unlike a live catalogue.
MAX_WORDS = 12

# What the model is told the slot is. "hhmm" beside a list of words would ask for two answers at once.
DESCRIPTION = "time of day"

# Every word this understands, folded, with the time it means. The hours are clock conventions rather than
# game events: only noon, midnight and the game's own "day" have one right answer.
TABLE = {
    # English
    "midnight": 0, "dawn": 600, "sunrise": 600, "morning": 800, "noon": 1200,
    "midday": 1200, "day": 1200, "daytime": 1200, "afternoon": 1500,
    "evening": 1900, "sunset": 2000, "dusk": 2000, "night": 2300,
    # German
    "mitternacht": 0, "morgengrauen": 600, "sonnenaufgang": 600, "morgens": 800,
    "morgen": 800, "vormittag": 1000, "mittag": 1200, "tag": 1200,
    "nachmittag": 1500, "abend": 1900, "abends": 1900, "sonnenuntergang": 2000,
    "nacht": 2300, "nachts": 2300,
    # Spanish
    "medianoche": 0, "amanecer": 600, "manana": 800, "mediodia": 1200, "dia": 1200,
    "tarde": 1500, "atardecer": 2000, "anochecer": 2000, "noche": 2300,
    # French
    "minuit": 0, "aube": 600, "matin": 800, "midi": 1200, "jour": 1200,
    "journee": 1200, "apresmidi": 1500, "soir": 1900, "soiree": 1900,
    "coucherdusoleil": 2000, "nuit": 2300,
}

ACCENTS = str.maketrans({
    "á": "a", "à": "a", "â": "a", "ä": "a", "ã": "a", "å": "a",
    "é": "e", "è": "e", "ê": "e", "ë": "e",
    "í": "i", "ì": "i", "î": "i", "ï": "i",
    "ó": "o", "ò": "o", "ô": "o", "ö": "o", "õ": "o",
    "ú": "u", "ù": "u", "û": "u", "ü": "u",
    "ñ": "n", "ç": "c", "ß": "ss",
})


def fold(value: str) -> str:
    """Letters and digits, lower case, without accents."""
    return "".join(c for c in str(value or "").lower().translate(ACCENTS) if c.isalnum())


def owns(label: str) -> bool:
    """True when this argument label names a time of day."""
    return "hhmm" in str(label or "").lower()


def has_clock_marker(text: str) -> bool:
    """Whether this text says it is a time rather than merely being a number that could be read as one."""
    folded = fold(text)
    if sum(c.isdigit() for c in folded) >= 3:
        return True
    if ":" in text or "." in text:
        return True
    suffix = folded.lstrip("0123456789")
    return suffix in ("am", "pm", "uhr") or suffix.startswith("h")


def readings(query: str) -> list[tuple[str, bool]]:
    """Every console reading the player's own words carry - "8am", "8 uhr", "20:00" - and whether it was marked.

    Read in pairs as well as singly, because the hour and its unit are usually two words.
    """
    found: list[tuple[str, bool]] = []
    parts = str(query or "").split()
    for at, part in enumerate(parts):
        pair = part + " " + parts[at + 1] if at + 1 < len(parts) else None
        for text in (part, pair):
            if text is None or fold(text) in TABLE:
                continue
            reading = token(text)
            if reading is not None:
                # Both spellings are reported, duplicates and all. "8" and "8 uhr" are the same reading and only
                # the second says it is a clock, so dropping the duplicate here would drop the marker with it.
                found.append((reading, has_clock_marker(text)))
    return found


def choices(query: str) -> list[str]:
    """A clock reading the player wrote, then the words their wording contains, then the core.

    The reading comes first because the word list otherwise pulls the answer away from a number that was
    there all along, and the core words are dropped entirely once the query carries a clock marker - offered
    beside "800", the model still answered midnight.
    """
    chosen: list[str] = []
    on_the_clock = False
    for reading, marker in readings(query):
        if len(chosen) >= MAX_WORDS:
            break
        on_the_clock = on_the_clock or marker
        if reading not in chosen:
            chosen.append(reading)
    folded = fold(query)

    # Longest first, so "nachmittag" is offered before the "mittag" inside it and "apresmidi" before "midi".
    if folded:
        for word in sorted((w for w in TABLE if w in folded), key=lambda w: (-len(w), w)):
            if len(chosen) >= MAX_WORDS:
                break
            if word not in chosen:
                chosen.append(word)

    # The core words are for a request that names no time at all. Offered beside a reading the player wrote,
    # they win it: given ["800", "midnight", "dawn", ...] the model answers midnight.
    if not on_the_clock:
        for word in CORE:
            if len(chosen) >= MAX_WORDS:
                break
            if word not in chosen:
                chosen.append(word)

    return chosen


def token(value) -> str | None:
    """The console token for a word or a clock reading, or None when this is not a time at all."""
    text = str(value if value is not None else "").strip()
    if not text:
        return None

    reading = _clock(text)
    if reading is not None:
        return str(reading)

    known = TABLE.get(fold(text))
    return None if known is None else str(known)


def _clock(text: str) -> int | None:
    """Digits, then an optional separator and minutes, then an optional am/pm.

    Three or four digits with no separator are the console's own form and stay as they are, so 1530 is half
    past three rather than fifteen hundred hours past midnight.
    """
    at, hours, digits = 0, 0, 0
    while at < len(text) and text[at].isdigit() and digits < 4:
        hours = hours * 10 + int(text[at])
        digits += 1
        at += 1

    if digits == 0:
        return None

    minutes, split = 0, False

    if at < len(text) and text[at] in ":.hH":
        separator = text[at]
        at += 1
        minute_digits = 0
        while at < len(text) and text[at].isdigit() and minute_digits < 2:
            minutes = minutes * 10 + int(text[at])
            minute_digits += 1
            at += 1
        # "8h" is eight o'clock; "8:" is a stray colon and the reading stops being one.
        if minute_digits == 0 and separator == ":":
            return None
        split = True
    elif digits >= 3:
        hours, minutes = divmod(hours, 100)
        split = True

    suffix = fold(text[at:])
    if suffix == "pm" and hours < 12:
        hours += 12
    elif suffix == "am" and hours == 12:
        hours = 0
    elif suffix and suffix not in ("am", "pm", "uhr", "h"):
        return None

    if hours > 24 or minutes > 59:
        return None
    if hours == 24 and minutes > 0:
        return None
    if not split and digits > 2:
        return None

    return (hours % 24) * 100 + minutes


def word_for(query: str, reading: str | None) -> str | None:
    """The table word in this query that means this reading, if the player used one."""
    if reading is None:
        return None
    folded = fold(query)
    for word, value in TABLE.items():
        if str(value) == reading and word in folded:
            return word
    return None


def recoverable(query: str, reading: str | None) -> bool:
    """True when the runtime could reach this reading from the player's own words.

    Either a word from the table, or a clock expression in the query itself - "8am", "8 uhr", "20:00". The
    pair is read as well as the single token because the hour and its unit are usually two words.
    """
    if reading is None:
        return False
    if word_for(query, reading) is not None:
        return True

    parts = str(query or "").split()
    for at, part in enumerate(parts):
        if token(part) == reading:
            return True
        if at + 1 < len(parts) and token(part + " " + parts[at + 1]) == reading:
            return True
    return False
